src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -1,45 +1,481 @@
+            self.installs.remove(&pkg_id);
+        }
+        Ok(())
+    }
+
+    fn package_for_bin(&self, bin_name: &str) -> Option<PackageId> {
+        self.installs
+            .iter()
+            .find(|(_, info)| info.bins.contains(bin_name))
+            .map(|(pkg_id, _)| *pkg_id)
+    }
+
+    fn mark_installed(
+        &mut self,
+        pkg: &Package,
+        bins: &BTreeSet<String>,
+        version_req: Option<String>,
+        opts: &CompileOptions<'_>,
+        target: String,
+        rustc: String,
+    ) {
+        // Remove bins from any other packages.
+        for info in &mut self.installs.values_mut() {
+            for bin in bins {
+                info.bins.remove(bin);
+            }
+        }
+        // Remove empty metadata lines. If only BTreeMap had `retain`.
+        let to_remove = self
+            .installs
+            .iter()
+            .filter_map(|(&p, info)| if info.bins.is_empty() { Some(p) } else { None })
+            .collect::<Vec<_>>();
+        for p in to_remove.iter() {
+            self.installs.remove(p);
+        }
+        // Add these bins.
+        if let Some(info) = self.installs.get_mut(&pkg.package_id()) {
+            for bin in bins {
+                info.bins.remove(bin);
+            }
+            info.version_req = version_req;
+            info.features = feature_set(&opts.features);
+            info.all_features = opts.all_features;
+            info.no_default_features = opts.no_default_features;
+            info.profile = profile_name(opts.build_config.release).to_string();
+            info.target = Some(target);
+            info.rustc = Some(rustc);
+        } else {
+            self.installs.insert(
+                pkg.package_id(),
+                InstallInfo {
+                    version_req,
+                    bins: bins.clone(),
+                    features: feature_set(&opts.features),
+                    all_features: opts.all_features,
+                    no_default_features: opts.no_default_features,
+                    profile: profile_name(opts.build_config.release).to_string(),
+                    target: Some(target),
+                    rustc: Some(rustc),
+                    other: BTreeMap::new(),
+                },
+            );
+        }
+    }
+
+    fn remove(&mut self, pkg_id: PackageId, bins: &BTreeSet<String>) {
+        let mut info_entry = match self.installs.entry(pkg_id) {
+            btree_map::Entry::Occupied(e) => e,
+            btree_map::Entry::Vacant(..) => panic!("v2 unexpected missing `{}`", pkg_id),
+        };
+
+        for bin in bins {
+            info_entry.get_mut().bins.remove(bin);
+        }
+        if info_entry.get().bins.is_empty() {
+            info_entry.remove();
+        }
+    }
+
+    fn save(&self, lock: &FileLock) -> CargoResult<()> {
+        let mut file = lock.file();
+        file.seek(SeekFrom::Start(0))?;
+        file.set_len(0)?;
+        serde_json::to_writer(file, self)?;
```

Comment:
FWIW we've has perf problems historically where `to_writer` on an unbuffered file does thousands of little one-character writes, so it may be best to just serialize to an in-memory string and then write it out?